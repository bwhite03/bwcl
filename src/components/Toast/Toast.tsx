import React from "react";
import { Default } from "../../utils";
import { toastManager, Event } from "../../core";
import { AnimationTypes, ToastProps } from "../../types";
import { Icons } from "../Icons";

const Toast: React.FC<ToastProps> = (props) => {
  const {
    content,
    toastId,
    type,
    theme = "light",
    position = "top-right",
    toastAnimation,
    toastShowIcon,
    toastClassName,
    toastBodyStyle,
  } = props;

  let { animation = "slide", showIcon = false } = props;

  if (toastAnimation) {
    animation = toastAnimation as AnimationTypes;
  }

  if (toastShowIcon === true || toastShowIcon === false) {
    showIcon = toastShowIcon;
  }

  const divAnimationClassName = `${Default.CSS_NAMESPACE}__${animation}-enter--${position}`;
  const classNames = `${Default.CSS_NAMESPACE}__toast ${Default.CSS_NAMESPACE}__toast--${type} ${divAnimationClassName} ${toastClassName}`;
  const maybeIcon = Icons[type as keyof typeof Icons];
  const iconProps = { theme, type };
  let Icon: React.ReactNode = maybeIcon && maybeIcon(iconProps);

  if (showIcon === false) {
    Icon = void 0;
  } else if (type === "info") {
    Icon = Icons.info(iconProps);
  } else if (type === "success") {
    Icon = Icons.success(iconProps);
  } else if (type === "error") {
    Icon = Icons.error(iconProps);
  } else if (type === "warning") {
    Icon = Icons.warning(iconProps);
  }

  const handleRemoveToast = () => {
    const toastDiv = document.getElementById(toastId as string);
    if (toastDiv) {
      toastDiv.classList.remove(
        `${Default.CSS_NAMESPACE}__${animation}-enter-${position}`
      );
      toastDiv.classList.add(
        `${Default.CSS_NAMESPACE}__${animation}-exit--${position}`
      );

      toastDiv.style.animationDuration = "1s";
      toastDiv.style.animationFillMode = "forwards";

      setTimeout(() => {
        toastManager.publish(Event.Clear, toastId);
      }, 1000);
    }
  };

  return (
    <div
      className={classNames}
      style={toastBodyStyle}
      id={toastId as string}
      onClick={handleRemoveToast}
      role="main"
    >
      {Icon && (
        <div className={`${Default.CSS_NAMESPACE}__toast-icon}`}>{Icon}</div>
      )}
      <div className={`${Default.CSS_NAMESPACE}__toast-message`}>{content}</div>
    </div>
  );
};

export default Toast;
