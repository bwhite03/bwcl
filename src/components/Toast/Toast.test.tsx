import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Toast from "./Toast";

describe("Toast", () => {
  test("it should render correctly", () => {
    const { container } = render(
      <Toast
        content={"Here is my toast"}
        toastId={"123"}
        type={"success"}
        position={"top-right"}
      />
    );

    expect(container).toMatchSnapshot();
  });

  test("should render without position", () => {
    const { container } = render(
      // @ts-ignore
      <Toast content={"Here is toast"} toastId={"123"} type={"success"} />
    );
    expect(container).toMatchSnapshot();
  });

  test("should show success icon", () => {
    const { container } = render(
      <Toast
        content={"Here is toast"}
        toastId={"123"}
        type={"success"}
        toastShowIcon={true}
        position="top-right"
        toastAnimation="slide"
      />
    );
    expect(container).toMatchSnapshot();
  });

  test("should show info icon", () => {
    const { container } = render(
      <Toast
        content={"Here is toast"}
        toastId={"123"}
        type={"info"}
        toastShowIcon={true}
        position="top-right"
        toastAnimation="slide"
      />
    );
    expect(container).toMatchSnapshot();
  });

  test("should show error icon", () => {
    const { container } = render(
      <Toast
        content={"Here is toast"}
        toastId={"123"}
        type={"error"}
        toastShowIcon={true}
        position="top-right"
        toastAnimation="slide"
      />
    );
    expect(container).toMatchSnapshot();
  });

  test("should show warning icon", () => {
    const { container } = render(
      <Toast
        content={"Here is toast"}
        toastId={"123"}
        type={"warning"}
        toastShowIcon={true}
        position="top-right"
        toastAnimation="slide"
      />
    );
    expect(container).toMatchSnapshot();
  });

  test("should handle removal of tosat", () => {
    const { container, getByRole } = render(
      <Toast
        content={"toast"}
        toastId={"123"}
        type={"warning"}
        position="top-right"
      />
    );

    const div = getByRole("main");
    fireEvent.click(div);
  });
});
