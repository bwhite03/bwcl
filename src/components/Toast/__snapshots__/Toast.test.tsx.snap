// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast it should render correctly 1`] = `
<div>
  <div
    class="Toastify__toast Toastify__toast--success Toastify__slide-enter--top-right undefined"
    id="123"
    role="main"
  >
    <div
      class="Toastify__toast-message"
    >
      Here is my toast
    </div>
  </div>
</div>
`;

exports[`Toast should render without position 1`] = `
<div>
  <div
    class="Toastify__toast Toastify__toast--success Toastify__slide-enter--top-right undefined"
    id="123"
    role="main"
  >
    <div
      class="Toastify__toast-message"
    >
      Here is toast
    </div>
  </div>
</div>
`;

exports[`Toast should show error icon 1`] = `
<div>
  <div
    class="Toastify__toast Toastify__toast--error Toastify__slide-enter--top-right undefined"
    id="123"
    role="main"
  >
    <div
      class="Toastify__toast-icon"
    >
      <svg
        fill-opacity="var(--toastify__toast-icon-color-error)"
        height="100%"
        viewBox="0 0 24 24"
        width="100%"
      />
    </div>
    <div
      class="Toastify__toast-message"
    >
      Here is toast
    </div>
  </div>
</div>
`;

exports[`Toast should show info icon 1`] = `
<div>
  <div
    class="Toastify__toast Toastify__toast--info Toastify__slide-enter--top-right undefined"
    id="123"
    role="main"
  >
    <div
      class="Toastify__toast-icon"
    >
      <svg
        fill-opacity="var(--toastify__toast-icon-color-info)"
        height="100%"
        viewBox="0 0 24 24"
        width="100%"
      />
    </div>
    <div
      class="Toastify__toast-message"
    >
      Here is toast
    </div>
  </div>
</div>
`;

exports[`Toast should show success icon 1`] = `
<div>
  <div
    class="Toastify__toast Toastify__toast--success Toastify__slide-enter--top-right undefined"
    id="123"
    role="main"
  >
    <div
      class="Toastify__toast-icon"
    >
      <svg
        fill-opacity="var(--toastify__toast-icon-color-success)"
        height="100%"
        viewBox="0 0 24 24"
        width="100%"
      />
    </div>
    <div
      class="Toastify__toast-message"
    >
      Here is toast
    </div>
  </div>
</div>
`;

exports[`Toast should show warning icon 1`] = `
<div>
  <div
    class="Toastify__toast Toastify__toast--warning Toastify__slide-enter--top-right undefined"
    id="123"
    role="main"
  >
    <div
      class="Toastify__toast-icon"
    >
      <svg
        fill-opacity="var(--toastify__toast-icon-color-warning)"
        height="100%"
        viewBox="0 0 24 24"
        width="100%"
      />
    </div>
    <div
      class="Toastify__toast-message"
    >
      Here is toast
    </div>
  </div>
</div>
`;
